pkg/app/piped/driftdetector/cloudrun/detector.go

```diff
@@ -13,3 +13,316 @@
+		appLister:         appLister,
+		gitClient:         gitClient,
+		stateGetter:       stateGetter,
+		reporter:          reporter,
+		appManifestsCache: appManifestsCache,
+		interval:          time.Minute,
+		config:            cfg,
+		secretDecrypter:   sd,
+		gitRepos:          make(map[string]git.Repo),
+		logger:            logger,
+	}
+}
+
+func (d *detector) Run(ctx context.Context) error {
+	d.logger.Info("start running drift detector for cloudrun applications")
+
+	ticker := time.NewTicker(d.interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			d.logger.Info("drift detector for cloudrun applications has been stopped")
+			return nil
+
+		case <-ticker.C:
+			d.check(ctx)
+		}
+	}
+}
+
+func (d *detector) ProviderName() string {
+	return d.provider.Name
+}
+
+func (d *detector) check(ctx context.Context) error {
+	appsByRepo := d.listGroupedApplication()
+
+	for repoID, apps := range appsByRepo {
+		gitRepo, ok := d.gitRepos[repoID]
+		if !ok {
+			// Clone repository for the first time.
+			gr, err := d.cloneGitRepository(ctx, repoID)
+			if err != nil {
+				d.logger.Error("failed to clone git repository",
+					zap.String("repo-id", repoID),
+					zap.Error(err),
+				)
+				continue
+			}
+			gitRepo = gr
+			d.gitRepos[repoID] = gitRepo
+		}
+
+		// Fetch the latest commit to compare the states.
+		branch := gitRepo.GetClonedBranch()
+		if err := gitRepo.Pull(ctx, branch); err != nil {
+			d.logger.Error("failed to pull repository branch",
+				zap.String("repo-id", repoID),
+				zap.Error(err),
+			)
+			continue
+		}
+
+		// Get the head commit of the repository.
+		headCommit, err := gitRepo.GetLatestCommit(ctx)
+		if err != nil {
+			d.logger.Error("failed to get head commit hash",
+				zap.String("repo-id", repoID),
+				zap.Error(err),
+			)
+			continue
+		}
+
+		// Start checking all applications in this repository.
+		for _, app := range apps {
+			if err := d.checkApplication(ctx, app, gitRepo, headCommit); err != nil {
+				d.logger.Error(fmt.Sprintf("failed to check application: %s", app.Id), zap.Error(err))
+			}
+		}
+	}
+	return nil
+}
+
+func (d *detector) cloneGitRepository(ctx context.Context, repoID string) (git.Repo, error) {
+	repoCfg, ok := d.config.GetRepository(repoID)
+	if !ok {
+		return nil, fmt.Errorf("repository %s was not found in piped configuration", repoID)
+	}
+	return d.gitClient.Clone(ctx, repoID, repoCfg.Remote, repoCfg.Branch, "")
+}
+
+// listGroupedApplication retrieves all applications those should be handled by this director
+// and then groups them by repoID.
+func (d *detector) listGroupedApplication() map[string][]*model.Application {
+	var (
+		apps = d.appLister.ListByCloudProvider(d.provider.Name)
+		m    = make(map[string][]*model.Application)
+	)
+	for _, app := range apps {
+		repoID := app.GitPath.Repo.Id
+		if _, ok := m[repoID]; !ok {
+			m[repoID] = []*model.Application{app}
+		} else {
+			m[repoID] = append(m[repoID], app)
+		}
```

Comment:
I'm guessing you don't have to initialize it anymore.

```suggestion
		m[repoID] = append(m[repoID], app)
```

Reply:
https://go.dev/play/p/ZenNfzUFt9Z

Reply:
You're right!
TYVM:)